File: scratch/clean_companions.py
import json
import os

def main():
    file_path = r"c:\Users\us mohamed\Desktop\مصحف\assets\data\companions.json"
    if not os.path.exists(file_path):
        print("Error: companions.json not found!")
        return

    try:
        with open(file_path, 'r', encoding='utf-8') as f:
            data = json.load(f)
        
        print(f"Total companions in database: {len(data)}")
        
        for comp in data:
            # Clean sources to strictly show Al-Isti'ab and Siyar A'lam al-Nubala
            comp["sources"] = "الاستيعاب في معرفة الأصحاب لابن عبد البر، سير أعلام النبلاء للذهبي"
            
            # Also clean individual sources inside sections if any
            # (lineage, islam, moments, virtues) to prevent referencing other books
            comp["lineage"] = comp["lineage"].split("(المصدر:")[0].strip() + " (المصدر: الاستيعاب لابن عبد البر)"
            comp["islam"] = comp["islam"].split("(المصدر:")[0].strip() + " (المصدر: سير أعلام النبلاء للذهبي)"
            comp["moments"] = comp["moments"].split("(المصدر:")[0].strip() + " (المصدر: سير أعلام النبلاء للذهبي)"
            comp["virtues"] = comp["virtues"].split("(المصدر:")[0].strip() + " (المصدر: سير أعلام النبلاء للذهبي)"
            
        with open(file_path, 'w', encoding='utf-8') as f:
            json.dump(data, f, ensure_ascii=False, indent=2)
            
        print("Successfully cleaned sources for all companions in companions.json.")
    except Exception as e:
        print(f"Error: {e}")

if __name__ == "__main__":
    main()

File: scratch/clean_companions.dart
import 'dart:convert';
import 'dart:io';

void main() {
  final file = File(r'c:\Users\us mohamed\Desktop\مصحف\assets\data\companions.json');
  if (!file.existsSync()) {
    print('Error: companions.json not found!');
    return;
  }

  try {
    final content = file.readAsStringSync();
    final List<dynamic> data = json.decode(content);

    print('Total companions in database: ${data.length}');

    for (var comp in data) {
      // Clean sources to strictly show Al-Isti'ab and Siyar A'lam al-Nubala
      comp['sources'] = 'الاستيعاب في معرفة الأصحاب لابن عبد البر، سير أعلام النبلاء للذهبي';

      // Clean individual sources inside sections if any
      // (lineage, islam, moments, virtues) to prevent referencing other books
      comp['lineage'] = cleanText(comp['lineage'] as String) + ' (المصدر: الاستيعاب لابن عبد البر)';
      comp['islam'] = cleanText(comp['islam'] as String) + ' (المصدر: سير أعلام النبلاء للذهبي)';
      comp['moments'] = cleanText(comp['moments'] as String) + ' (المصدر: سير أعلام النبلاء للذهبي)';
      comp['virtues'] = cleanText(comp['virtues'] as String) + ' (المصدر: سير أعلام النبلاء للذهبي)';
    }

    final encoder = JsonEncoder.withIndent('  ');
    file.writeAsStringSync(encoder.convert(data));
    print('Successfully cleaned sources for all companions in companions.json via Dart.');
  } catch (e) {
    print('Error: $e');
  }
}

String cleanText(String text) {
  if (text.contains('(المصدر:')) {
    return text.split('(المصدر:')[0].trim();
  }
  return text.trim();
}
